cloud/tasks/logging/journald_logger.go

```diff
@@ -42,23 +47,104 @@ func getPriority(level Level) journal.Priority {
 
 ////////////////////////////////////////////////////////////////////////////////
 
-func getFields(fields ...log.Field) map[string]string {
+type rawField struct {
+	key   string
+	value string
+}
+
+func getField(field log.Field) rawField {
+	// The variable name must be in uppercase and consist only of characters,
+	// numbers and underscores, and may not begin with an underscore:
+	// https://www.freedesktop.org/software/systemd/man/sd_journal_print.html
+	return rawField{
+		key:   strings.ToUpper(field.Key()),
+		value: fmt.Sprintf("%v", field.Any()),
+	}
+}
+
+// Takes the latest value for each key
+func getFieldsMap(fields ...log.Field) map[string]string {
 	if len(fields) == 0 {
 		return nil
 	}
 
 	vars := make(map[string]string)
 	for _, f := range fields {
-		// The variable name must be in uppercase and consist only of characters,
-		// numbers and underscores, and may not begin with an underscore:
-		// https://www.freedesktop.org/software/systemd/man/sd_journal_print.html
-		key := strings.ToUpper(f.Key())
-		vars[key] = fmt.Sprintf("%v", f.Any())
+		rawField := getField(f)
+		vars[rawField.key] = rawField.value
 	}
 
 	return vars
 }
 
+// Takes the latest value for each key
+// Preserves order of fields
+func deduplicateFields(
+	fields ...rawField,
+) (deduplicated []rawField) {
+
+	fieldsMap := make(map[string]string)
+	for _, f := range fields {
+		fieldsMap[f.key] = f.value
+	}
+
+	for _, f := range fields {
+		if value, ok := fieldsMap[f.key]; ok {
+			deduplicated = append(
+				deduplicated,
+				rawField{key: f.key, value: value},
+			)
+			delete(fieldsMap, f.key)
+		}
+	}
+
+	return
+}
+
+func serializeFields(fields []rawField) (fieldsStr string) {
+	for idx, f := range fields {
+		fieldsStr += f.key + ": " + f.value
+		if idx+1 < len(fields) {
+			fieldsStr += ", "
+		}
+	}
+	return
+}
+
+type journaldLoggerFileds struct {
+	journaldFields map[string]string
+	messageFields  []rawField
+	errorField     string
+}
+
+func newJournaldLoggerFileds(fields ...log.Field) (ff journaldLoggerFileds) {
+	fmt.Printf("newJournaldLoggerFileds: len = %d, %v\n", len(fields), fields)
+	ff.journaldFields = make(map[string]string)
+
+	var messageFieldsAll []rawField
+
+	for _, f := range fields {
+		if f.Key() == syslogIdentifierKey {
```

Comment:
We need to choose which fields to wriite to journald fields (would not be seen in journalctl without -o flag) and which fields to write to the message itself.

Currently we write syslog identifier, idempotency key, request id and operation id to journald fields.

In this pr syslog identifier is written to journald fields, and all other fields are written to the message.

It it ok?

Reply:
I think we can duplicate them. journalctl would allow us to perform filtering and tracing.